// paths/index.js
/**
 * This is HEAVILY inspired by the patterns found
 * in the paths module ejected from react-scripts (CRA)
 */

const config = require('../boring-config');
const path = require('path');

const proj_dir = path.normalize(config.get('paths.proj_dir', process.cwd()))
const base_app_path_key = 'paths.base_app_path'
const base_app_path = path.normalize(config.get(base_app_path_key, 'src'))
const boring_dir = path.normalize(__dirname + '/../../../')

function getProjPath(key, defaultVal) {
  return path.normalize(proj_dir +'/'+ config.get(key, defaultVal));
}

function getAppDirPath(key, defaultVal) {
  return path.normalize(getProjPath(base_app_path_key, base_app_path) + '/'+ config.get(key , defaultVal));
}

function getBoringPath(key, defaultVal) {
  return path.normalize(boring_dir + '/' + config.get(key, defaultVal));
}

module.exports = {  

  proj_dir,
  boring_dir,
  base_app_path,
  app_dir: getProjPath(base_app_path_key, base_app_path),

  server_endpoints: getAppDirPath('paths.server_endpoints', '/server/endpoints'),
  server_middleware: getAppDirPath('paths.server_middleware', '/server/middleware'),
  server_hooks: getAppDirPath('paths.server_hooks', '/server/hooks'),

  boring_endpoints: getBoringPath('paths.boring_endpoints', '/src/lib/init-endpoints/core-endpoints'),
  boring_middleware: getBoringPath('paths.boring_middleware', '/src/lib/init-middleware/core-middleware'),
  boring_hooks: getBoringPath('paths.boring_hooks', '/src/lib/init-hooks/core-hooks'),

  boring_webpack_dev_config: getBoringPath('paths.boring_webpack_dev_config', '/config/webpack.config.dev.js'),
  boring_webpack_prod_config: getBoringPath('paths.boring_webpack_prod_config', '/config/webpack.config.prod.js'),

  app_config: getProjPath('paths.app_config', '/config'),
  asset_manifest: getProjPath('paths.asset_manifest', '/build/asset-manifest.json') 
}
